core/embed/lib/error_handling.h

```diff
@@ -20,6 +20,151 @@
 #ifndef LIB_ERROR_HANDLING_H
 #define LIB_ERROR_HANDLING_H
 
+// Suppresses the intellisense error in VSCode
+#ifndef __FILE_NAME__
+#define __FILE_NAME__ __FILE__
+#endif
+
+// Build status code from any 16-bit value.
+// This macro is used for generic status codes.
+#define TS_BUILD(code) ((code) | (~(code) << 16))
+
+// Status codes
+//
+// Status code is hardened against fault injections
+// by storing the negated value in the upper 16 bits.
+typedef enum {
+  TS_OK = TS_BUILD(0),
+  TS_ERROR = TS_BUILD(1),          // Generic error
+  TS_ERROR_BUSY = TS_BUILD(2),     // Busy
+  TS_ERROR_TIMEOUT = TS_BUILD(3),  // Timeout
+  TS_ERROR_NOTINIT = TS_BUILD(4),  // Not initialized
+  TS_ERROR_ARG = TS_BUILD(4),      // Invalid argument
+  TS_ERROR_IO = TS_BUILD(5),       // I/O error
+} ts_t;
+
+// Check status code consistency and returns its value.
+// If invalid status code is detected, it will call `__fatal_error()`.
+#define _ts_checked(status)                                       \
+  ({                                                              \
+    ts_t _checked_status = (status);                              \
+    if ((_checked_status & 0xFFFF) != (~_checked_status >> 16)) { \
+      __fatal_error("ts_check() error", __FILE_NAME__, __LINE__); \
+    }                                                             \
+    _checked_status;                                              \
+  })
+
+// Returns `true` if status code is `TS_OK`
+#define ts_ok(status) (_ts_checked(status) == TS_OK)
+
+// Returns `true` if status code is NOT `TS_OK`
+#define ts_error(status) (_ts_checked(status) != TS_OK)
+
+// ----------------------------------------------------
+// verify_init(), verify_status() and verify_xxx() macros define
+// a simple error handling mechanism
+//
+// Example:
+//
+// ts_t my_function(int arg) {
+//   // initialize verify mechanism
+//   verify_init();
+//
+//   // check arguments
+//   verify_arg(arg > 0);
+//
+//   ts_t status;
+//
+//   // verify success
+//   status = some_function();
+//   verify_ok(status);
+//
+//   // verify condition
+//   verify(another_function() != 0, TS_ERROR_IO);
+//
+//  error:
+//
+//   // clean up code comes here
+//
+//   return verify_status();
+// }
+
+// Declares a status variable and initializes it to `TS_OK`.
+// This variable is used to store the status
+#define verify_init() __attribute__((unused)) ts_t __status = TS_OK;
+
+// Returns the current status.
+#define verify_status() (__status)
+
+// Jumps to `error` label if status is not `TS_OK`.
+#define verify_ok(status)    \
+  do {                       \
+    ts_t _status = status;   \
+    if (ts_error(_status)) { \
+      __status = _status;    \
+      goto error;            \
```

Comment:
I don't like the idea of using macros that perform jumps or return. It looks like a function, but has an unexpected side effect for someone who is not intimately familiar with it.

Reply:
I understand, it's a lot about getting used to it. People are always confused when they see it for the first time. However, this way of coding is very compact, and my experience is that it ends up being very easy to read.

Reply:
In past I've used more weird name instead of `verify_xxx() ` so the side effect was expected;-) Maybe different name could help.

Reply:
Another perspective is that using this macro enables us to log all errors in the debug or emulator build.

Reply:
I don't doubt that it is very compact and I am sure that if all devs in the team would work with it daily, it would be very easy to read and they would find nothing unexpected about the side effects. However, I think that most of the development now gets done in Python and in Rust. So if a dev needs to work with the C code only occasionally, then they have to relearn the idioms again each time. Furthermore, since we are open-source, anyone who wants to look at our code would have to familiarize themselves with this idiom before being able to understand what's going on.

One example of macros with side effects that we have in our code are those in `legacy/firmware/fsm.c`: `CHECK_PIN`, `CHECK_PIN_UNCACHED` etc. I feel it's always a real pain to touch the code that uses these macros. It would be easier for me if they were inlined.

Reply:
I understand your point of view. Unexpected side effects are almost always inappropriate - `CHECK_PIN()` and its counterparts are good examples. But I don't believe that `verify_xxx()` falls into this category since it's a cross-cutting concept used throughout the entire codebase (hopefully in the future:-), and the side effect is quite apparent and local.

Our code is open source and should be readable for anybody without special 'training'. However, I think it already contains a lot of complexity, so it's not true that it's accessible to everybody for a long time. 

But yes, I understand your concerns.

Reply:
If we do decide to go ahead with this idiom, then we should at least use uppercase for the macros that have side effects to emphasize to the reader that they are macros. `verify_ok()` -> `TS_CHECK()`